File: Data Structure & Algorithm/Template/Binary Tree/Binary Tree Traversal/Level-order/Level-order.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	queue := []*TreeNode{}

	if root == nil {
		return res
	}

	queue = append(queue, root)
	for len(queue) > 0 {
		size := len(queue)
		temp := []int{}
		for i := 0; i < size; i++ {
			current := queue[0]
			queue = queue[1:]
			temp = append(temp, current.Val)
			if current.Left != nil {
				queue = append(queue, current.Left)
			}
			if current.Right != nil {
				queue = append(queue, current.Right)
			}
		}
		res = append(res, temp)
	}

	return res
}

// test
func main() {
	root := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 4,
			},
			Right: &TreeNode{
				Val: 5,
			},
		},
		Right: &TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val: 6,
				Left: &TreeNode{
					Val: 7,
				},
			},
		},
	}

	res := levelOrder(root)
	fmt.Println(res)
}

File: Data Structure & Algorithm/Models/Tree/Binary Tree/Binary Tree Traversal/Level-order/Level-order.py
from typing import Any, List


class TreeNode:

    def __init__(self, val, left=None, right=None) -> None:
        self.val: Any = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


def levelOrder(root: TreeNode) -> List[Any]:
    res = []
    queue: List[TreeNode] = []

    if root == None:
        return res

    queue.append(root)
    while len(queue) > 0:
        size = len(queue)
        temp = []
        for _ in range(size):
            current = queue.pop(0)
            temp.append(current.val)
            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        res.append(temp)

    return res


# test
root = TreeNode(1)
node2 = TreeNode(2)
node3 = TreeNode(3)
node4 = TreeNode(4)
node5 = TreeNode(5)
node6 = TreeNode(6)
node7 = TreeNode(7)

root.left = node2
root.right = node3
node2.left = node4
node2.right = node5
node3.left = node6
node6.left = node7

res = levelOrder(root)
print(res)